### PR TITLE
Preserve live byte arguments across local reloads

### PR DESCRIPTION
**Pull request by jonathanpeppers**
🤖 ## Summary

Fix an operand-lifetime miscompile in ordinary managed, extern, and default-path NESLib calls. When a runtime byte remains on the IL evaluation stack while a later argument reloads a local or parameter, select operands for the existing typed snapshot/reload pass. Follow producer dependencies through conversions and arithmetic. Extern methods use the ordinary cc65 stack-call classification, including supported Boolean operands; managed Boolean parameters remain rejected. Pure parameter forwarding retains its existing no-snapshot path. The writer and snapshot selector now share the default-path `rect_overlap` / `sprite_overlap` target classification.

For `Operations.Store(second: Operations.Read(2), first: Operations.Read(1))` with an unused default byte argument, frontend IL is already correct. Native emission saves the live first argument in `TEMP`, but the call does not restore it and pushes the second value twice. Parameter reloads have the same defect. Typed snapshots preserve identity and evaluation order without named-argument special cases, source rewrites, new APIs, flags, or ABI changes. Existing width, alias, branch, interrupt, native-code and frame guards remain intact.

The frozen baseline is official c8fb1f2ff72a4abac0522894d3d6b6987a307f96 from successful main CI 35007539845. Raw and formatting-only controls produce CLR `01 02 02` versus native `02 02 02`. The latest candidate uses identical C# and decoded input IL and produces `01 02 02` for both. This proves the current-main defect, not when it was introduced; the original consumer test only covered CLR execution. Standalone repro main grows from 71 to 98 bytes, executes 91 instructions, and ends with software stack `$0800` and hardware stack `$FD` at the external stop breakpoint.

## Validation at f304f0e

- **41 emitted-execution cases and one managed-Boolean diagnostic regression** cover frozen source, named/default and positional/saved-local shapes, nested side effects, distinct edge bytes, physical argument/default slots, stack wrap/balance, forwarding, signed/narrowed results, local/parameter-derived later expressions, `ldarg.s`, caller parameter identity, pure forwarding without extra snapshots, native Booleans, mixed-width A:X tails, and real collision builtins. Initial 17 cases failed on unchanged main; all 14 fixed review-boundary cases were reproduced before editing.
- Full solution/CLI/decompiler Debug and Release builds pass. **2,700 compiler tests and 68 analyzer tests per configuration**, with strict unchanged sample snapshots.
- [CI run 35018830391](https://github.com/jonathanpeppers/dotnes/actions/runs/35018830391) is green: Linux builds/tests/packages/samples and Windows/macOS emulator smoke tests.
- **1,070 latest-package desktop/ROM checks** pass against the official baseline, including unchanged hello ROM and native assets.
- Independent consumer proof was renewed for this exact package: fresh full solution/ROM Rebuild, **723/723 tests**, no failures/skips; raw/control repro parity; 25 exact pixel checkpoints including both restarts; 10,000-frame runtime verification.

### Consumer ROM and runtime proof

The latest game ROM remains **byte-identical to the raw official-c8 output and all earlier candidates**: 229,392 bytes, SHA256 `3D18D2E6B7E340C509C2C75BC807B84CD075E25C8677B6755D47B7273F73D413`. The review fixes change no game instructions or budgets.

| Measure | Latest result |
| --- | --- |
| Program / main / free bytes | 15,143 / 10,227 / 1,227; capacity 16,370 |
| 10,000-frame run | first acknowledgment frame 11; 9,989 acknowledgments; 9,990 publications |
| Maximum foreground cycles | 28,301, strictly below 29,780 |
| Native path / software stack | 2,104 cycles / `$0800` |
| Exact pixel checkpoints | 25, zero differing pixels |

Relative to the consumer's separately preserved normalized-main baseline, raw-source totals are program -92 bytes, main -65 bytes, foreground +106 cycles. Relative to the previous raw candidate, all deltas are zero.

## Latest premerge candidate provenance

HEAD and nuspec repository commit: `f304f0e259d6a367e4e79357484b2680d3623544`.

Package: `dotnes.0.3.3-reordered.f304f0e.4.nupkg`.
SHA256: `CDE1B6BDFE47D2DA6184487E5AFA8FE87A3515BAFC2DD5356983FABAE842B2B1`.

Identical build/ref/lib compiler DLLs: SHA256 `650750BCA0B5D4D75DD172913C509073D5D6DFE6DB8DE26B47D013C627A0E15B`; assembly version `0.3.3.0`; informational version `0.3.3+f304f0e259d6a367e4e79357484b2680d3623544`.

An explicit Release compiler Rebuild preceded packing with isolated SDK 10.0.401. Local packaging bypassed only the repository Directory.Build.targets sample-cache cleanup because a previous sample build held its cached DLL open; no compiler options changed. Unmodified CI builds are green. The compatible isolated SDK avoids the installed 10.0.300 SDK's inability to load the existing Roslyn 5.9 analyzer. Full provenance and independent repro/game artifacts were delivered to the consumer; earlier evidence is preserved.

## Review disposition and known limitation

All three actual inline review threads were individually answered and resolved. The latest review summary raised two additional findings without inline threads: default-path collision helpers are fixed in f304f0e, and the confirmed general scalar struct-field lifetime defect is explicitly deferred to #627. That case needs currently absent struct-field type and declaring-address analysis; merely broadening the reload predicate is insufficient. It remains a known correctness gap, not a case fixed or validated by this PR. Both summary findings received separate explanatory replies.

The consumer restored its official default package graph and freshly rebuilt official outputs afterward. **Final adoption requires a successful official main-CI package after human merge, not this local candidate. No merge or release performed.**

### CODE REVIEW

File: src/dotnes.tasks/Utilities/Transpiler.ExpressionValues.cs [hunk 1]
```diff
@@ -53,6 +53,23 @@ bool HasPreservedPadResult(int producer)
...
+                // after a local reload. Snapshot the operands in evaluation order.
+                for (int argument = 0; argument < inputs.Length - 1; argument++)
+                {
+                    int producer = inputs[argument];
+                    if (producer >= 0 && scalar[producer]
+                        && types[producer] is PrimitiveTypeCode.Byte or PrimitiveTypeCode.SByte
+                        && instructions[producer].GetLdlocIndex() == null
+                        && instructions[producer].GetLdcValue() == null
+                        && inputs.Skip(argument + 1).Any(p => p > producer
+                            && instructions[p].GetLdlocIndex() != null))
```
[the remark this answers is not in the document]
Comment (jonathanpeppers) [reply src/dotnes.tasks/Utilities/Transpiler.ExpressionValues.cs]: 🤖 Fixed in 57366e69eb3b3508fc65f8d13f331b492edba29e. The later argument's producer dependencies are now inspected for local reloads after the live earlier value, rather than checking only the argument root. Added emitted-execution regressions for `(byte)(saved + 1)`, `(byte)(saved ^ 255)`, and `(byte)(saved << 1)`. All three fail before this change and pass afterward, checking both values, exactly two Read calls, and balanced hardware/software stacks. Full Debug/Release compiler suites pass (2,689 each), with unchanged sample snapshots.

File: src/dotnes.tasks/Utilities/Transpiler.ExpressionValues.cs [hunk 2]
```diff
@@ -53,6 +53,23 @@ bool HasPreservedPadResult(int producer)
...
+            if (instructions[i].OpCode == ILOpCode.Call && instructions[i].String is string target
+                && (UserMethods.ContainsKey(target) || ExternMethods.ContainsKey(target)))
+            {
+                // Ordinary calls cannot recover an earlier runtime byte from TEMP
+                // after a local reload. Snapshot the operands in evaluation order.
+                for (int argument = 0; argument < inputs.Length - 1; argument++)
+                {
+                    int producer = inputs[argument];
+                    if (producer >= 0 && scalar[producer]
+                        && types[producer] is PrimitiveTypeCode.Byte or PrimitiveTypeCode.SByte
```
[the remark this answers is not in the document]
Comment (jonathanpeppers) [reply src/dotnes.tasks/Utilities/Transpiler.ExpressionValues.cs]: 🤖 Fixed the supported extern case in 57366e69eb3b3508fc65f8d13f331b492edba29e by including Boolean in typed operand preservation. A real emitted native callee now receives `[1, 0]` instead of the reproduced `[0, 0]`, with both side effects and stack balance checked. One scope correction: managed Boolean parameters are not currently supported; ConfigureNumericTypes explicitly rejects them. Added a separate diagnostic regression preserving that existing restriction rather than introducing a new managed ABI.

**Comment by jonathanpeppers** [comment]
🤖 Final verification at `57366e69eb3b3508fc65f8d13f331b492edba29e`: CI [35013404772](https://github.com/jonathanpeppers/dotnes/actions/runs/35013404772) is green, including both platform smoke tests. Both concrete review findings were reproduced, fixed, individually answered, and resolved.

Local Debug/Release validation passes 2,689 compiler tests and 68 analyzer tests per configuration, with unchanged sample snapshots. The final hash-identified package passes 1,070 independent desktop/ROM package checks.

The consumer independently repeated its fresh full Rebuild and all 723 tests against `0.3.3-reordered.57366e6.2`. Raw and frozen-control repros now match CLR `01 02 02`. Its game ROM remains byte-identical to the official-c8 raw build, with all 25 exact pixel checkpoints and 10,000-frame cadence/size/native gates unchanged. Full provenance and metrics are recorded in the PR description.

Ready for human review. No merge or release performed; final consumer adoption still requires a successful official main-CI package after merge, not this local validation package.

File: src/dotnes.tasks/Utilities/Transpiler.ExpressionValues.cs [hunk 3]
```diff
@@ -50,9 +50,29 @@ bool HasPreservedPadResult(int producer)
                     || i.OpCode == ILOpCode.Call && i.String is "pad_poll" or "pad_trigger");
         }
 
+        bool ReloadsLocalAfter(int value, int earlier) => value > earlier
+            && (instructions[value].GetLdlocIndex() != null
+                || analysis.Inputs[value].Any(input => ReloadsLocalAfter(input, earlier)));
```
[the remark this answers is not in the document]
Comment (jonathanpeppers) [reply src/dotnes.tasks/Utilities/Transpiler.ExpressionValues.cs]: 🤖 Fixed in e4d95b0e44a9d86ca4d8ab880e795362658533c9. The dependency walk now recognizes all parameter-load forms through NumericArgIndex, including loads inside later arithmetic expressions. Eight execution regressions reproduced the failure before the fix; they now check distinct edge-byte values, exactly-once Read calls, unchanged caller parameters, ldarg.s forwarding, and balanced hardware/software stacks. A ninth regression ensures ordinary pure-parameter forwarding keeps its existing no-snapshot path. Full Debug and Release builds pass, with 2,698 compiler tests and 68 analyzer tests per configuration and unchanged sample snapshots.

**Comment by jonathanpeppers** [comment]
🤖 Addressing the default-path NESLib finding in review 5215274460: fixed in f304f0e259d6a367e4e79357484b2680d3623544. Snapshot selection now shares the existing `rect_overlap` / `sprite_overlap` target classifier with the writer instead of silently excluding those cc65 calls.

Two emitted-execution regressions reproduce the issue before the fix: `sprite_overlap(Read(1), saved, 1, 64, 8)` and `rect_overlap(Read(1), saved, 8, 8, 1, 64, 8, 8)`, with `saved = 64`, both returned false instead of true. Both now execute the real builtins correctly, call Read once, and restore software/hardware stacks. Debug and Release full solution builds and 2,700 compiler + 68 analyzer tests per configuration pass, as do 1,070 latest-package checks. No unchanged sample snapshots were updated.

**Comment by jonathanpeppers** [comment]
🤖 Addressing the scalar struct-field finding in review 5215274460: confirmed and explicitly deferred to #627, not claimed fixed by this PR. Native execution of `Store(Read(1), p.X)` with `p.X = 255` produces `[0, 255]` instead of `[1, 255]`.

Merely treating `ldfld` as a clobbering reload is insufficient here. The existing expression pass has type information for closure fields, but not general user-struct fields; selecting the earlier byte leaves an untyped later operand in the closed spill group. Resolving field type and declaring struct/address identity safely is broader work than this focused local/parameter reload fix. Field operands retain an unqualified name, so guessing a byte width or inferring types by field name would be unsafe. The follow-up records the reproducer, root cause, and execution/width/identity requirements. This known struct-field limitation remains open.

This finding was only in the review summary, so it has no inline thread to resolve. All three actual inline threads have individual replies and are resolved.

**Comment by jonathanpeppers** [comment]
🤖 Review iteration is complete for the focused scope at f304f0e259d6a367e4e79357484b2680d3623544. All three inline threads have individual replies and are resolved; both additional summary findings have separate dispositions. The confirmed general scalar struct-field defect remains explicitly deferred to #627.

Latest CI 35018830391 is green, including samples and both Windows/macOS emulator smoke jobs. Independent consumer validation of package SHA256 `CDE1B6BDFE47D2DA6184487E5AFA8FE87A3515BAFC2DD5356983FABAE842B2B1` is complete: fresh full Rebuild, 723/723 tests, exact frozen raw/control repro fixed, unchanged game ROM bytes, all 25 pixel checkpoints exact, and unchanged 10,000-frame cadence/native/size budgets. The PR description records the full compiler/package provenance and validation.

Ready for human review with the documented limitation. No merge, release, or candidate adoption performed; final consumer adoption still requires a successful official main-CI package after human merge.
